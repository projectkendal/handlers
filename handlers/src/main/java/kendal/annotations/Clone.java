package kendal.annotations;

import java.lang.annotation.ElementType;
import java.lang.annotation.Target;

/**
 * Applied to a method creates new method with name from {@link #methodName()} attribute and list of parameters same as original method.
 * Generated method calls the original method, passes the result to an instance of {@link #transformer()}
 * and returns transformed value. If {@link #methodName()} is not specified, clone with default name is created.
 * Default name is: originalName + "Clone"
 * Return type of generated method is extracted from declaration of {@link #transformer()} class.
 */
@Target(ElementType.METHOD)
public @interface Clone {

    Class<? extends Transformer> transformer();

    /**
     * Java does not allow putting {@link java.lang.annotation.Annotation}, which is the base type for all annotations, as annotation member.
     * Let's just use this field as if it was defined.
     * We remove onMethod parameter from AST after processing @Clone usage.
     * /
//    Annotation[] onMethod() default {};

    /**
     * Name of the generated method.
     * Defaults to original method name + "Clone" suffix
     * */
    String methodName() default "";

    interface Transformer<T,R> {
        R transform(T input);
    }
}
