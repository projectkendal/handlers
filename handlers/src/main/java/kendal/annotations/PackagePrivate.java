package kendal.annotations;

import java.lang.annotation.ElementType;
import java.lang.annotation.Target;

@Target(value= ElementType.PARAMETER)
public @interface PackagePrivate {

    boolean makeFinal() default true;
}