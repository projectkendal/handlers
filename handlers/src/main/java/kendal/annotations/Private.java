package kendal.annotations;

import java.lang.annotation.ElementType;
import java.lang.annotation.Target;

@Target(value= ElementType.PARAMETER)
public @interface Private {
}