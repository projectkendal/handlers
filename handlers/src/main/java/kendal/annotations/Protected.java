package kendal.annotations;

import java.lang.annotation.ElementType;
import java.lang.annotation.Target;

@Target({ElementType.PARAMETER, ElementType.ANNOTATION_TYPE})
public @interface Protected {

    boolean makeFinal() default true;
}