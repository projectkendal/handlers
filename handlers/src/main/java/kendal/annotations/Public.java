package kendal.annotations;

import java.lang.annotation.ElementType;
import java.lang.annotation.Target;

@Target(value= ElementType.PARAMETER)
public @interface Public {

    boolean makeFinal() default true;
}