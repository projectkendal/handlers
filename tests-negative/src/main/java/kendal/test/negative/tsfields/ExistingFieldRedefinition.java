package kendal.test.negative.tsfields;

import kendal.annotations.Private;

public class ExistingFieldRedefinition {

    Object field;

    // This kind of construct indicates that user does not understand how @Private works;
    // field is generated automatically, so having it defined manually as well misses the point of using annotation.
    // We signal an error.
    ExistingFieldRedefinition(@Private Object field) {
    }
}
