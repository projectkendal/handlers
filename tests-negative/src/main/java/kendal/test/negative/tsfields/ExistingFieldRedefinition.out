- compiler.err.proc.messager: [Kendal] Auto generated field was already defined manually in this class!
1 error